File: vsa/examples/01-todo-list-ts/src/contexts/tasks/complete-task/CompleteTaskHandler.ts
import { CompleteTaskCommand } from './CompleteTaskCommand';
import { TaskCompletedEvent } from './TaskCompletedEvent';
import { EventStore } from '../../../infrastructure/EventStore';
import { TaskAggregate } from '../create-task/TaskAggregate';

/**
 * Handler for CompleteTask command
 * 
 * Business Rules:
 * - Task must exist
 * - Task must not be already completed
 * - Task must not be deleted
 */
export class CompleteTaskHandler {
  constructor(private eventStore: EventStore) {}

  async handle(command: CompleteTaskCommand): Promise<void> {
    // Load task aggregate from events
    const events = await this.eventStore.getEvents(command.id);
    
    if (events.length === 0) {
      throw new Error(`Task with ID ${command.id} not found`);
    }

    // Rebuild aggregate state
    const aggregate = new TaskAggregate();
    for (const event of events) {
      this.applyEvent(aggregate, event);
    }

    // Business rule validation
    if (aggregate.isCompleted()) {
      throw new Error('Task is already completed');
    }

    if (aggregate.isDeleted()) {
      throw new Error('Cannot complete a deleted task');
    }

    // Create event
    const event: TaskCompletedEvent = {
      id: command.id,
      completedAt: new Date(),
    };

    // Store event
    await this.eventStore.appendEvent(command.id, 'TaskCompleted', event);
  }

  private applyEvent(aggregate: TaskAggregate, event: any): void {
    switch (event.type) {
      case 'TaskCreated':
        aggregate.applyTaskCreated(event.data);
        break;
      case 'TaskCompleted':
        aggregate.applyTaskCompleted(event.data);
        break;
      case 'TaskDeleted':
        aggregate.applyTaskDeleted(event.data);
        break;
    }
  }
}

File: vsa/examples/05-todo-list-py/src/contexts/tasks/complete-task/CompleteTaskHandler.py
"""Handler for CompleteTaskCommand"""

from .CompleteTaskCommand import CompleteTaskCommand
from .CompleteTaskEvent import CompleteTaskEvent
from event_sourcing import Repository

class CompleteTaskHandler:
    """Handler for CompleteTaskCommand

    This handler processes the command, applies business logic,
    creates events, and persists them to the event store.
    """

    def __init__(self, repository: Repository):
        self.repository = repository

    async def handle(self, command: CompleteTaskCommand) -> None:
        """Process the command and emit events"""
        # TODO: Add validation logic

        # Create event
        event = CompleteTaskEvent(
            id=command.id,
        )

        # TODO: Persist to event store
        # await self.event_store.append(stream_name, [event])
